README.md
# QbtClient

A Ruby gem to access qBittorrent's [WebUI](https://github.com/qbittorrent/qBittorrent/wiki/WebUI-API-Documentation)

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'qbt_client'
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install qbt_client

## Usage

Instantiate the client

```ruby
require 'qtb_client'

ip   = 'http://127.0.0.1' # Protocol is required.
port = 8083
user = 'admin'
pass = 'abcabc'  # Min length for password is 6 chars

client = QtbClient::WebUI.new(ip, port, user, pass)
```

Call methods on the client

```ruby
# Get list of torrents:
torrents = client.torrent_list


torrent_properties = {}

# Using each torrent's hash, get the torrents properties:
torrents.each do |t|
  hash = t['hash']

  torrent_properties[hash] = client.properties hash

  # Get the torrent's trackers too
  torrent_properties[hash]['trackers'] = client.trackers hash
end
```

## Testing

To run the tests, you'll need to have qBittorrent installed locally, WebUI
turned on, and the credentials set to:

- user: admin
- pass: abcabc

The tests assume that qBittorrent is running at 127.0.0.1, port 8083.

From the root project dir, run:

    $ rspec spec/

## Contributing

1. Fork it ( https://github.com/jmcaffee/qbt_client/fork )
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Create your tests
4. Commit your changes (`git commit -am 'Add some feature'`)
5. Push to the branch (`git push origin my-new-feature`)
6. Create a new Pull Request



CHANGES.md
# qbt_client CHANGELOG

## v1.0.0

Update client to work with qBitTorrent 3.3.4 WebUI changes

- Login method has changed from digest authentication to cookie based token
  - NOTE: minimum password length is now 6 chars
- `/json/` url paths have changed to `/query/`
- `/command/pauseall` has changed to `/command/pauseAll` (all paths are case sensitive)
- `/command/resumeall` has changed to `/command/resumeAll`
- `/command/getTorrentDlLimit` changed to `/command/getTorrentsDlLimit`
  - `hash` option changed to `hashes` option
- `/command/setTorrentDlLimit` changed to `/command/setTorrentsDlLimit`
  - `hash` option changed to `hashes` option
- `/command/getTorrentUpLimit` changed to `/command/getTorrentsUpLimit`
  - `hash` option changed to `hashes` option
- `/command/setTorrentUpLimit` changed to `/command/setTorrentsUpLimit`
  - `hash` option changed to `hashes` option
- Turn on network debug output when `DEBUG` env var is set


## v0.1.0

Initial release


spec/spec_helper.rb
$LOAD_PATH.unshift File.expand_path('../../lib', __FILE__)
require 'qbt_client'
require 'digest'

def print_response res
  puts 'response:'
  puts res.inspect
end

def test_ip
  'http://127.0.0.1'
end

def test_port
  8083
end

def test_user
  'admin'
end

def test_pass
  'abcabc'
end

def test_torrent_url
  "http://www.slackware.com/torrents/slackware64-14.1-install-dvd.torrent"
end

def test_torrent_name
  "slackware64-14.1-iso"
end

def test_torrent_url2
  "http://torrent.unix-ag.uni-kl.de/torrents/KNOPPIX_V7.4.2DVD-2014-09-28-EN.torrent"
end

def test_torrent_name2
  "KNOPPIX_V7.4.2DVD-2014-09-28-EN"
end

def get_torrent_info client, torrent_hash, key
  info = {}
  client.torrent_list.each do |t|
    if t["hash"] == torrent_hash
      info = t
    end
  end
  info[key]
end

def hash_from_torrent_name client, name
  torrents = client.torrent_list
  hash = nil

  torrents.each do |t|
    if t["name"] == name
      hash = t["hash"]
    end
  end
  hash
end

def delete_torrent_with_hash client, hash
  unless hash.nil? or hash.empty?
    client.delete_torrent_and_data hash
  end
end

def delete_torrent_with_name client, name
  unless name.nil? or name.empty?
    hash = hash_from_torrent_name client, name
    delete_torrent_with_hash client, hash
  end
end

def download_torrent name, url
  client = QbtClient::WebUI.new(test_ip, test_port, test_user, test_pass)

  # If we're already downloading the torrent, delete it
  delete_torrent_with_name client, name

  # Start the download...
  client.download url
  sleep 2

  hash = hash_from_torrent_name client, name

  # Return the info.
  [hash, name]
end

def download_and_pause_torrent name, url
  client = QbtClient::WebUI.new(test_ip, test_port, test_user, test_pass)

  # If we're already downloading the torrent, just pause it and return its data
  hash = hash_from_torrent_name client, name
  unless hash.nil?
    client.pause hash
    return [hash, name]
  end

  # Torrent not found, start the download...
  client.download url
  sleep 2

  hash = hash_from_torrent_name client, name

  # Pause the download...
  client.pause hash

  # Return the info.
  [hash, name]
end

def given_a_downloading_torrent
  download_torrent test_torrent_name, test_torrent_url
end

def given_another_downloading_torrent
  download_torrent test_torrent_name2, test_torrent_url2
end

def given_a_paused_downloading_torrent
  download_and_pause_torrent test_torrent_name, test_torrent_url
end

def given_another_paused_downloading_torrent
  download_and_pause_torrent test_torrent_name2, test_torrent_url2
end

def torrent_exists? client, hash
  return false if (client.properties(hash)).nil?
  true
end

def torrent_with_name_exists? client, name
  hash = hash_from_torrent_name client, name

  return false if (client.properties(hash)).nil?
  true
end

def md5 str
  Digest::MD5.hexdigest str
end

def enable_queueing client, enable
  client.set_preferences({ "queueing_enabled" => enable })
end

def queueing_enabled? client
  prefs = client.preferences
  prefs['queueing_enabled']
end

def delete_all_torrents
  client = QbtClient::WebUI.new(test_ip, test_port, test_user, test_pass)
  torrents = client.torrent_list
  hash = nil

  torrents.each do |t|
    client.delete_torrent_and_data t["hash"]
  end
end



spec/lib/qbt_client_spec.rb
require 'spec_helper'

describe QbtClient do
  it 'has a version number' do
    expect(QbtClient::VERSION).not_to be nil
  end
end


spec/lib/qbittorrent_spec.rb
require 'spec_helper'

describe Qbittorrent do
  it 'has a version number' do
    expect(Qbittorrent::VERSION).not_to be nil
  end

  it 'does something useful' do
    expect(false).to eq(true)
  end
end


lib/qbt_client/client.rb
##############################################################################
# File::    client.rb
# Purpose:: Web UI client for qBittorrent.
#
# Author::    Jeff McAffee 2015-02-07
# Copyright:: Copyright (c) 2015, kTech Systems LLC. All rights reserved.
# Website::   http://ktechsystems.com
##############################################################################

require 'json/pure'
require 'httparty'
require 'digest'


module QbtClient

  class Client
    include HTTParty

    #debug_output $stdout

    ###
    # constructor
    #
    def initialize(ip, port, user, pass)
      @ip         = ip
      @port       = port
      @user       = user
      @pass       = pass

      self.class.digest_auth(user, pass)
      self.class.base_uri "#{ip}:#{port}"
    end

    ###
    # Get array of all torrents
    #
    def torrent_list
      self.class.format :json
      self.class.get('/json/torrents').parsed_response
    end

    def torrent_data torrent_hash
      torrents = torrent_list

      torrents.each do |t|
        if t["hash"] == torrent_hash
          return t
        end
      end
    end

    def properties torrent_hash
      self.class.format :json
      self.class.get('/json/propertiesGeneral/' + torrent_hash).parsed_response
    end

    def trackers torrent_hash
      self.class.format :json
      self.class.get('/json/propertiesTrackers/' + torrent_hash).parsed_response
    end

    def contents torrent_hash
      self.class.format :json
      self.class.get('/json/propertiesFiles/' + torrent_hash).parsed_response
    end

    def transfer_info
      self.class.format :json
      self.class.get('/json/transferInfo').parsed_response
    end

    def preferences
      self.class.format :json
      self.class.get('/json/preferences').parsed_response
    end

    def preferences= pref_hash
      pref_hash = Hash(pref_hash)
      options = {
        body: "json=#{pref_hash.to_json}"
      }

      self.class.post('/command/setPreferences', options)
    end

    def pause torrent_hash
      options = {
        body: "hash=#{torrent_hash}"
      }

      self.class.post('/command/pause', options)
    end

    def pause_all
      self.class.post('/command/pauseall')
    end

    def resume torrent_hash
      options = {
        body: "hash=#{torrent_hash}"
      }

      self.class.post('/command/resume', options)
    end

    def resume_all
      self.class.post('/command/resumeall')
    end

    def download urls
      urls = Array(urls)
      urls = urls.join('%0A')

      options = {
        body: "urls=#{urls}"
      }

      self.class.post('/command/download', options)
    end

    def delete_torrent_and_data torrent_hashes
      torrent_hashes = Array(torrent_hashes)
      torrent_hashes = torrent_hashes.join('|')

      options = {
        body: "hashes=#{torrent_hashes}"
      }

      self.class.post('/command/deletePerm', options)
    end

    def delete torrent_hashes
      torrent_hashes = Array(torrent_hashes)
      torrent_hashes = torrent_hashes.join('|')

      options = {
        body: "hashes=#{torrent_hashes}"
      }

      self.class.post('/command/delete', options)
    end

    def recheck torrent_hash
      options = {
        body: "hash=#{torrent_hash}"
      }

      self.class.post('/command/recheck', options)
    end

    def increase_priority torrent_hashes
      torrent_hashes = Array(torrent_hashes)
      torrent_hashes = torrent_hashes.join('|')

      options = {
        body: "hashes=#{torrent_hashes}"
      }

      self.class.post('/command/increasePrio', options)
    end

    def decrease_priority torrent_hashes
      torrent_hashes = Array(torrent_hashes)
      torrent_hashes = torrent_hashes.join('|')

      options = {
        body: "hashes=#{torrent_hashes}"
      }

      self.class.post('/command/decreasePrio', options)
    end

    def maximize_priority torrent_hashes
      torrent_hashes = Array(torrent_hashes)
      torrent_hashes = torrent_hashes.join('|')

      options = {
        body: "hashes=#{torrent_hashes}"
      }

      self.class.post('/command/topPrio', options)
    end

    def minimize_priority torrent_hashes
      torrent_hashes = Array(torrent_hashes)
      torrent_hashes = torrent_hashes.join('|')

      options = {
        body: "hashes=#{torrent_hashes}"
      }

      self.class.post('/command/bottomPrio', options)
    end

  private

    def md5 str
      Digest::MD5.hexdigest str
    end
  end
end



lib/qbt_client/web_ui.rb
##############################################################################
# File::    web_ui.rb
# Purpose:: Web UI client for qBittorrent.
#
# Author::    Jeff McAffee 2015-02-07
# Copyright:: Copyright (c) 2015, kTech Systems LLC. All rights reserved.
# Website::   http://ktechsystems.com
##############################################################################

require 'json/pure'
require 'httparty'
require 'digest'


module QbtClient

  class WebUI
    include HTTParty

    if ENV["DEBUG"]
      debug_output $stdout
    end

    ###
    # constructor
    #
    def initialize(ip, port, user, pass)
      @ip         = ip
      @port       = port
      @user       = user
      @pass       = pass
      @sid        = nil

      #self.class.digest_auth(user, pass)
      self.class.base_uri "#{ip}:#{port}"
      authenticate
      self.class.cookies.add_cookies(@sid)
    end


    ###
    # Authenticate with the server
    #
    # Login with username and password.
    # Store returned SID cookie value used as auth token for later calls.
    #
    def authenticate
      options = {
        body: "username=#{@user}&password=#{@pass}"
      }

      # Have to clear out the cookies or the old SID gets sent while requesting
      # the new SID (and it fails).
      self.class.cookies.clear

      res = self.class.post('/login', options)
      if res.success?
        token = res.headers["Set-Cookie"]
        raise "Login failed" if token.nil?

        token = token.split(";")[0]
        #token = token.split("SID=")[1]
        @sid = token
      end
    end

    ###
    # Get array of all torrents
    #
    # Example response:
    #    [
    #      {
    #          "dlspeed"=>"3.1 MiB/s",
    #          "eta"=>"9m",
    #          "hash"=>"156b69b8643bd11849a5d8f2122e13fbb61bd041",
    #          "name"=>"slackware64-14.1-iso",
    #          "num_leechs"=>"1 (14)",
    #          "num_seeds"=>"97 (270)",
    #          "priority"=>"*",
    #          "progress"=>0.172291,
    #          "ratio"=>"0.0",
    #          "size"=>"2.2 GiB",
    #          "state"=>"downloading",
    #          "upspeed"=>"0 B/s"
    #      },
    #      {
    #        "dlspeed"=>"1.8 KiB/s",
    #        "eta"=>"28d 1h",
    #        "hash"=>"1fe5775d32d3e58e48b3a96dd2883c5250882cda",
    #        "name"=>"Grimm.S04E12.720p.HDTV.X264-DIMENSION.mkv",
    #        "num_leechs"=>"7 (471)",
    #        "num_seeds"=>"15 (1866)",
    #        "priority"=>"*",
    #        "progress"=>1.53669e-07,
    #        "ratio"=>"0.0",
    #        "size"=>"825.4 MiB",
    #        "state"=>"downloading",
    #        "upspeed"=>"0 B/s"
    #      }
    #    ]
    #
    def torrent_list
      self.class.format :json
      self.class.get('/query/torrents').parsed_response
    end

    def torrent_data torrent_hash
      torrents = torrent_list

      torrents.each do |t|
        if t["hash"] == torrent_hash
          return t
        end
      end
    end

    ###
    # Get properties of a torrent (different data than what's returned
    # in #torrent_list).
    #
    # Example response:
    #    {
    #      "comment"=>"Visit us: https://eztv.ch/ - Bitcoin: 1EZTVaGQ6UsjYJ9fwqGnd45oZ6HGT7WKZd",
    #      "creation_date"=>"Friday, February 6, 2015 8:01:22 PM MST",
    #      "dl_limit"=>"∞",
    #      "nb_connections"=>"0 (100 max)",
    #      "piece_size"=>"512.0 KiB",
    #      "save_path"=>"/home/jeff/Downloads/",
    #      "share_ratio"=>"0.0",
    #      "time_elapsed"=>"< 1m",
    #      "total_downloaded"=>"646.8 KiB (657.8 KiB this session)",
    #      "total_uploaded"=>"0 B (0 B this session)",
    #      "total_wasted"=>"428 B",
    #      "up_limit"=>"∞"
    #    }
    #
    def properties torrent_hash
      self.class.format :json
      self.class.get('/query/propertiesGeneral/' + torrent_hash).parsed_response
    end

    ###
    # Get tracker data for a torrent
    #
    # Example response:
    #    [
    #      {
    #        "msg"=>"",
    #        "num_peers"=>"0",
    #        "status"=>"Working",
    #        "url"=>"udp://open.demonii.com:1337"},
    #      {
    #        "msg"=>"",
    #        "num_peers"=>"0",
    #        "status"=>"Not contacted yet",
    #        "url"=>"udp://tracker.coppersurfer.tk:6969"},
    #      {
    #        "msg"=>"",
    #        "num_peers"=>"0",
    #        "status"=>"Not contacted yet",
    #        "url"=>"udp://tracker.leechers-paradise.org:6969"},
    #      {
    #        "msg"=>"",
    #        "num_peers"=>"0",
    #        "status"=>"Not contacted yet",
    #        "url"=>"udp://exodus.desync.com:6969"}
    #    ]
    #
    def trackers torrent_hash
      self.class.format :json
      self.class.get('/query/propertiesTrackers/' + torrent_hash).parsed_response
    end

    ###
    # Add one or more trackers to a torrent
    #
    # If passing mulitple urls, pass them as an array.
    #
    def add_trackers torrent_hash, urls
      urls = Array(urls)
      # Ampersands in urls must be escaped.
      urls = urls.map { |url| url.gsub('&', '%26') }
      urls = urls.join('%0A')

      options = {
        body: "hash=#{torrent_hash}&urls=#{urls}"
      }

      self.class.post('/command/addTrackers', options)
    end

    ###
    # Get torrent contents (files data)
    #
    # Example response:
    #    [
    #      {
    #        "is_seed"=>false,
    #        "name"=>"Grimm.S04E12.720p.HDTV.X264-DIMENSION.mkv",
    #        "priority"=>1,
    #        "progress"=>0.0,
    #        "size"=>"825.4 MiB"
    #      }
    #    ]
    #
    def contents torrent_hash
      self.class.format :json
      self.class.get('/query/propertiesFiles/' + torrent_hash).parsed_response
    end

    ###
    # Get application transfer info
    #
    # Example response:
    #    {
    #      "dl_info"=>"D: 0 B/s/s - T: 657.8 KiB",
    #      "up_info"=>"U: 0 B/s/s - T: 0 B"
    #    }
    #
    def transfer_info
      self.class.format :json
      self.class.get('/query/transferInfo').parsed_response
    end

    ###
    # Get application preferences (options)
    #
    # Example response:
    #    {
    #      "alt_dl_limit"=>10,
    #      "alt_up_limit"=>10,
    #      "anonymous_mode"=>false,
    #      "autorun_enabled"=>false,
    #      "autorun_program"=>"",
    #      "bypass_local_auth"=>false,
    #      "dht"=>true,
    #      "dhtSameAsBT"=>true,
    #      "dht_port"=>6881,
    #      "dl_limit"=>-1,
    #      "dont_count_slow_torrents"=>false,
    #      "download_in_scan_dirs"=>[],
    #      "dyndns_domain"=>"changeme.dyndns.org",
    #      "dyndns_enabled"=>false,
    #      "dyndns_password"=>"",
    #      "dyndns_service"=>0,
    #      "dyndns_username"=>"",
    #      "enable_utp"=>true,
    #      "encryption"=>0,
    #      "export_dir"=>"",
    #      "export_dir_enabled"=>false,
    #      "incomplete_files_ext"=>false,
    #      "ip_filter_enabled"=>false,
    #      "ip_filter_path"=>"",
    #      "limit_tcp_overhead"=>false,
    #      "limit_utp_rate"=>true,
    #      "listen_port"=>6881,
    #      "locale"=>"en_US",
    #      "lsd"=>true,
    #      "mail_notification_auth_enabled"=>false,
    #      "mail_notification_email"=>"",
    #      "mail_notification_enabled"=>false,
    #      "mail_notification_password"=>"",
    #      "mail_notification_smtp"=>"smtp.changeme.com",
    #      "mail_notification_ssl_enabled"=>false,
    #      "mail_notification_username"=>"",
    #      "max_active_downloads"=>3,
    #      "max_active_torrents"=>5,
    #      "max_active_uploads"=>3,
    #      "max_connec"=>500,
    #      "max_connec_per_torrent"=>100,
    #      "max_uploads_per_torrent"=>4,
    #      "pex"=>true,
    #      "preallocate_all"=>false,
    #      "proxy_auth_enabled"=>false,
    #      "proxy_ip"=>"0.0.0.0",
    #      "proxy_password"=>"",
    #      "proxy_peer_connections"=>false,
    #      "proxy_port"=>8080,
    #      "proxy_type"=>-1,
    #      "proxy_username"=>"",
    #      "queueing_enabled"=>false,
    #      "save_path"=>"/home/jeff/Downloads",
    #      "scan_dirs"=>[],
    #      "schedule_from_hour"=>8,
    #      "schedule_from_min"=>0,
    #      "schedule_to_hour"=>20,
    #      "schedule_to_min"=>0,
    #      "scheduler_days"=>0,
    #      "scheduler_enabled"=>false,
    #      "ssl_cert"=>"",
    #      "ssl_key"=>"",
    #      "temp_path"=>"/home/jeff/Downloads/temp",
    #      "temp_path_enabled"=>false,
    #      "up_limit"=>50,
    #      "upnp"=>true,
    #      "use_https"=>false,
    #      "web_ui_password"=>"ae150cdc82b40c4373d2e15e0ffe8f67",
    #      "web_ui_port"=>8083,
    #      "web_ui_username"=>"admin"
    #    }
    #
    def preferences
      self.class.format :json
      self.class.get('/query/preferences').parsed_response
    end

    ###
    # Set application preferences
    #
    # Note: When setting password, pass it as plain text.
    # You can send only the key/value pairs you want to change (in a hash),
    # rather than the entire set of data.
    #
    def set_preferences pref_hash
      pref_hash = Hash(pref_hash)
      options = {
        body: "json=#{pref_hash.to_json}"
      }

      self.class.post('/command/setPreferences', options)
    end

    ###
    # Pause a torrent
    #
    def pause torrent_hash
      options = {
        body: "hash=#{torrent_hash}"
      }

      self.class.post('/command/pause', options)
    end

    ###
    # Pause all torrents
    #
    def pause_all
      self.class.post('/command/pauseAll')
    end

    ###
    # Resume downloading/seeding of a torrent
    #
    def resume torrent_hash
      options = {
        body: "hash=#{torrent_hash}"
      }

      self.class.post('/command/resume', options)
    end

    ###
    # Resume downloading/seeding of all torrents
    #
    def resume_all
      self.class.post('/command/resumeAll')
    end

    ###
    # Begin downloading one or more torrents.
    #
    # If passing mulitple urls, pass them as an array.
    #
    def download urls
      urls = Array(urls)
      urls = urls.join('%0A')

      options = {
        body: "urls=#{urls}"
      }

      self.class.post('/command/download', options)
    end

    ###
    # Delete one or more torrents AND THEIR DATA
    #
    # If passing multiple torrent hashes, pass them as an array.
    #
    def delete_torrent_and_data torrent_hashes
      torrent_hashes = Array(torrent_hashes)
      torrent_hashes = torrent_hashes.join('|')

      options = {
        body: "hashes=#{torrent_hashes}"
      }

      self.class.post('/command/deletePerm', options)
    end

    ###
    # Delete one or more torrents (doesn't delete their data)
    #
    # If passing multiple torrent hashes, pass them as an array.
    #
    def delete torrent_hashes
      torrent_hashes = Array(torrent_hashes)
      torrent_hashes = torrent_hashes.join('|')

      options = {
        body: "hashes=#{torrent_hashes}"
      }

      self.class.post('/command/delete', options)
    end

    ###
    # Recheck a torrent
    #
    def recheck torrent_hash
      options = {
        body: "hash=#{torrent_hash}"
      }

      self.class.post('/command/recheck', options)
    end

    ###
    # Increase the priority of one or more torrents
    #
    # If passing multiple torrent hashes, pass them as an array.
    # Note: This does nothing unless queueing has been enabled
    # via preferences.
    #
    def increase_priority torrent_hashes
      torrent_hashes = Array(torrent_hashes)
      torrent_hashes = torrent_hashes.join('|')

      options = {
        body: "hashes=#{torrent_hashes}"
      }

      self.class.post('/command/increasePrio', options)
    end

    ###
    # Decrease the priority of one or more torrents
    #
    # If passing multiple torrent hashes, pass them as an array.
    # Note: This does nothing unless queueing has been enabled
    # via preferences.
    #
    def decrease_priority torrent_hashes
      torrent_hashes = Array(torrent_hashes)
      torrent_hashes = torrent_hashes.join('|')

      options = {
        body: "hashes=#{torrent_hashes}"
      }

      self.class.post('/command/decreasePrio', options)
    end

    ###
    # Increase the priority of one or more torrents to the maximum value
    #
    # If passing multiple torrent hashes, pass them as an array.
    # Note: This does nothing unless queueing has been enabled
    # via preferences.
    #
    def maximize_priority torrent_hashes
      torrent_hashes = Array(torrent_hashes)
      torrent_hashes = torrent_hashes.join('|')

      options = {
        body: "hashes=#{torrent_hashes}"
      }

      self.class.post('/command/topPrio', options)
    end

    ###
    # Decrease the priority of one or more torrents to the minimum value
    #
    # If passing multiple torrent hashes, pass them as an array.
    # Note: This does nothing unless queueing has been enabled
    # via preferences.
    #
    def minimize_priority torrent_hashes
      torrent_hashes = Array(torrent_hashes)
      torrent_hashes = torrent_hashes.join('|')

      options = {
        body: "hashes=#{torrent_hashes}"
      }

      self.class.post('/command/bottomPrio', options)
    end

    ###
    # Set the download priority of a file within a torrent
    #
    # file_id is a 0 based position of the file within the torrent
    #
    def set_file_priority torrent_hash, file_id, priority
      query = ["hash=#{torrent_hash}", "id=#{file_id}", "priority=#{priority}"]

      options = {
        body: query.join('&')
      }

      self.class.post('/command/setFilePrio', options)
    end

    ###
    # Get the application's global download limit
    #
    # A limit of 0 means unlimited.
    #
    # Returns an integer (bytes)
    #
    def global_download_limit
      self.class.format :json
      self.class.post('/command/getGlobalDlLimit').parsed_response
    end

    ###
    # Set the application's global download limit
    #
    # A limit of 0 means unlimited.
    #
    # limit: integer (bytes)
    #
    def set_global_download_limit limit
      query = "limit=#{limit}"

      options = {
        body: query
      }

      self.class.post('/command/setGlobalDlLimit', options)
    end

    ###
    # Get the application's global upload limit
    #
    # A limit of 0 means unlimited.
    #
    # Returns an integer (bytes)
    #
    def global_upload_limit
      self.class.format :json
      self.class.post('/command/getGlobalUpLimit').parsed_response
    end

    ###
    # Set the application's global upload limit
    #
    # A limit of 0 means unlimited.
    #
    # limit: integer (bytes)
    #
    def set_global_upload_limit limit
      query = "limit=#{limit}"

      options = {
        body: query
      }

      self.class.post('/command/setGlobalUpLimit', options)
    end

    ###
    # Get a torrent's download limit
    #
    # A limit of 0 means unlimited.
    #
    # Returns an integer (bytes)
    #
    def download_limit torrent_hash
      self.class.format :json

      options = {
        body: "hashes=#{torrent_hash}"
      }

      self.class
        .post('/command/getTorrentsDlLimit', options)
        .parsed_response[torrent_hash]
    end

    ###
    # Set a torrent's download limit
    #
    # A limit of 0 means unlimited.
    #
    # torrent_hash: string
    # limit: integer (bytes)
    #
    def set_download_limit torrent_hash, limit
      query = ["hashes=#{torrent_hash}", "limit=#{limit}"]

      options = {
        body: query.join('&')
      }

      self.class.post('/command/setTorrentsDlLimit', options)
    end

    ###
    # Get a torrent's upload limit
    #
    # A limit of 0 means unlimited.
    #
    # Returns an integer (bytes)
    #
    def upload_limit torrent_hash
      self.class.format :json

      options = {
        body: "hashes=#{torrent_hash}"
      }

      self.class
        .post('/command/getTorrentsUpLimit', options)
        .parsed_response[torrent_hash]
    end

    ###
    # Set a torrent's upload limit
    #
    # A limit of 0 means unlimited.
    #
    # torrent_hash: string
    # limit: integer (bytes)
    #
    def set_upload_limit torrent_hash, limit
      query = ["hashes=#{torrent_hash}", "limit=#{limit}"]

      options = {
        body: query.join('&')
      }

      self.class.post('/command/setTorrentsUpLimit', options)
    end

  private

    def md5 str
      Digest::MD5.hexdigest str
    end
  end
end

